target/linux/ar71xx/files/arch/mips/ath79/mach-wnr2000-v5.c

```diff
@@ -0,0 +1,212 @@
+#include <asm/mach-ath79/ath79_spi_platform.h>
+#include <asm/mach-ath79/ar71xx_regs.h>
+
+#include "common.h"
+#include "dev-eth.h"
+#include "dev-gpio-buttons.h"
+#include "dev-leds-gpio.h"
+#include "dev-m25p80.h"
+#include "dev-spi.h"
+#include "dev-usb.h"
+#include "dev-wmac.h"
+#include "machtypes.h"
+
+#define WNR2000V5_GPIO_LED_STATUS_GREEN 21
+#define WNR2000V5_GPIO_LED_STATUS_AMBER 20
+#define WNR2000V5_GPIO_LED_DATA_GREEN   25
+#define WNR2000V5_GPIO_LED_DATA_AMBER   19
+#define WNR2000V5_GPIO_LED_WLAN         18
+
+#define WNR2000V5_GPIO_BTN_RFKILL 0
+#define WNR2000V5_GPIO_BTN_RESET  1
+#define WNR2000V5_GPIO_BTN_WPS    2
+
+#define WNR2000V5_KEYS_POLL_INTERVAL	20	/* msecs */
+#define WNR2000V5_KEYS_DEBOUNCE_INTERVAL	(4 * WNR2000V5_KEYS_POLL_INTERVAL)
+
+#define WNR2000V5_SPI_GPIO_RESET 14
+#define WNR2000V5_SPI_GPIO_MOSI 15
+#define WNR2000V5_SPI_GPIO_CLK 16
+#define WNR2000V5_74X164_GPIO_BASE QCA953X_GPIO_COUNT // The 74x614 muxer adds 8 new gpio after the regular 18 QCA953X gpios
+
+
+static struct gpio_led wnr2000v5_leds_gpio[] __initdata = {
+	{
+		.name		= "netgear:green:status",
+		.gpio		= WNR2000V5_GPIO_LED_STATUS_GREEN,
+		.active_low	= 1,
+		.default_trigger = "default-on",
+	},
+	{
+		.name		= "netgear:amber:status",
+		.gpio		= WNR2000V5_GPIO_LED_STATUS_AMBER,
+		.active_low	= 1,
+	},
+	{
+		.name		= "netgear:green:wlan",
+		.gpio		= WNR2000V5_GPIO_LED_WLAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "netgear:green:wan",
+		.gpio		= WNR2000V5_GPIO_LED_DATA_GREEN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "netgear:amber:wan",
+		.gpio		= WNR2000V5_GPIO_LED_DATA_AMBER,
+		.active_low	= 1,
+	},
+};
+
+static struct gpio_keys_button wnr2000v5_gpio_keys[] __initdata = {
+	{
+		.desc		= "WPS button",
+		.type		= EV_KEY,
+		.code		= KEY_WPS_BUTTON,
+		.debounce_interval = WNR2000V5_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= WNR2000V5_GPIO_BTN_WPS,
+		.active_low	= 1,
+	},
+	{
+		.desc		= "rfkill",
+		.type		= EV_KEY,
+		.code		= KEY_RFKILL,
+		.debounce_interval = WNR2000V5_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= WNR2000V5_GPIO_BTN_RFKILL,
+		.active_low	= 1,
+	},
+	{
+		.desc		= "reset",
+		.type		= EV_KEY,
+		.code		= KEY_RESTART,
+		.debounce_interval = WNR2000V5_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= WNR2000V5_GPIO_BTN_RESET,
+		.active_low	= 1,
+	},
+};
+
+static void __init wnr2000v5_gpio_led_setup(void)
+{
+	ath79_gpio_direction_select(WNR2000V5_GPIO_LED_STATUS_GREEN, true);
+	ath79_gpio_direction_select(WNR2000V5_GPIO_LED_STATUS_AMBER, true);
+	ath79_gpio_direction_select(WNR2000V5_GPIO_LED_WLAN, true);
+	ath79_gpio_direction_select(WNR2000V5_GPIO_LED_DATA_GREEN, true);
+	ath79_gpio_direction_select(WNR2000V5_GPIO_LED_DATA_AMBER, true);
+
+	ath79_register_leds_gpio(-1, ARRAY_SIZE(wnr2000v5_leds_gpio),
+			wnr2000v5_leds_gpio);
+
+	ath79_register_gpio_keys_polled(-1, WNR2000V5_KEYS_POLL_INTERVAL,
+			ARRAY_SIZE(wnr2000v5_gpio_keys),
+			wnr2000v5_gpio_keys);
+}
+
+static u8 wnr2000v5_ssr_initdata[] __initdata = {
+	0xdf,
```

Comment:
wat

Reply:
It's a bitfield, frankly I'm not sure why I initialized it without BIT(5) but i can't test it anymore and it worked like that, so I'm not too willing to change it.